lib/tar.ml

```diff
@@ -412,41 +400,32 @@ module Header = struct
          - the <keyword> cannot contain an equals sign
          - the <length> is the number of octets of the record, including \n
         *)
-      let find buffer char =
-        let rec loop i =
-          if i = Cstruct.length buffer
-          then None
-          else if Cstruct.get_char buffer i = char
-          then Some i
-          else loop (i + 1)
-        in
-        loop 0
+      let find start char =
+        try Ok (String.index_from c start char)
+        with Not_found -> Error (`Unmarshal "Failed to decode pax extended header record")
       in
-      let rec loop remaining =
-        if Cstruct.length remaining = 0
+      let slen = String.length c in
+      let rec loop idx =
+        if idx = slen - 1
         then Ok []
         else begin
           (* Find the space, then decode the length *)
-          match find remaining ' ' with
-          | None -> Error (`Unmarshal "Failed to decode pax extended header record")
-          | Some i ->
-            let length = int_of_string @@ Cstruct.to_string @@ Cstruct.sub remaining 0 i in
-            let record = Cstruct.sub remaining 0 length in
-            let remaining = Cstruct.shift remaining length in
-            begin match find record '=' with
-            | None -> Error (`Unmarshal "Failed to decode pax extended header record")
-            | Some j ->
-              let keyword = Cstruct.to_string @@ Cstruct.sub record (i + 1) (j - i - 1) in
-              let v = Cstruct.to_string @@ Cstruct.sub record (j + 1) (Cstruct.length record - j - 2) in
-              let* rem = loop remaining in
-              Ok ((keyword, v) :: rem)
-            end
+          let* i = find idx ' ' in
+          let* length =
+            try Ok (int_of_string (String.sub c idx (i - idx))) with
+              Failure _ -> Error (`Unmarshal "Failed to decode pax extended header record")
+          in
+          let* j = find i '=' in
+          let keyword = String.sub c (i + 1) (j - i - 1)
+          and v = String.sub c (j + 1) (length - j - i - 1)
```

Comment:
The offset `j` includes `i`. it leads to an error. Another note, the loop is not tailcall... I have a patch but I'm not sure about the fix where I must ignore the trailing `\n` for values.